File: 02-进阶能力/code/day10/structured_output_demo.py
"""
structured_output_demo.py — 三种结构化输出方法对比

方法1: Prompt 约束（手工法）
方法2: JSON Mode（API 原生）
方法3: Function Calling（最强约束）

用法：
  python3 structured_output_demo.py
  对比三种方法的输出差异
"""

import json
import os
import sys
import requests


# ─── 读取 API Key ─────────────────────────────
def get_api_key():
    key = os.environ.get("DEEPSEEK_API_KEY", "")
    if not key:
        try:
            with open(os.path.expanduser("~/.hermes/auth.json")) as f:
                auth = json.load(f)
            pool = auth.get("credential_pool", {}).get("deepseek", [])
            if pool:
                key = pool[0].get("access_token", "")
        except Exception:
            pass
    return key


API_KEY = get_api_key()
API_URL = "https://api.deepseek.com/v1/chat/completions"


def call_llm(messages, **kwargs):
    """调用 DeepSeek API"""
    payload = {
        "model": "deepseek-chat",
        "messages": messages,
        "temperature": 0,
    }
    payload.update(kwargs)

    resp = requests.post(
        API_URL,
        headers={"Authorization": f"Bearer {API_KEY}", "Content-Type": "application/json"},
        json=payload,
        timeout=30,
    )
    return resp.json()["choices"][0]["message"]


# ═══════════════════════════════════════════════
# 方法1: Prompt 约束
# ═══════════════════════════════════════════════
def method1_prompt(text: str) -> dict:
    """通过 Prompt 让模型返回 JSON"""
    prompt = f"""从以下文本中提取信息，只返回 JSON（不要加任何其他文字）：

文本：{text}

要求的 JSON 格式：
{{
    "name": "姓名",
    "age": "年龄（数字）",
    "job": "职业",
    "city": "所在城市",
    "skills": ["技能列表"]
}}"""

    msg = call_llm([{"role": "user", "content": prompt}])
    content = msg["content"].strip()

    # 清理可能的 markdown 包裹
    if "```" in content:
        content = content.split("```")[1]
        if content.startswith("json"):
            content = content[4:]

    # 提取 JSON 部分
    start = content.find("{")
    end = content.rfind("}") + 1
    return json.loads(content[start:end]) if start >= 0 else {"error": "未找到 JSON"}


# ═══════════════════════════════════════════════
# 方法2: JSON Mode
# ═══════════════════════════════════════════════
def method2_json_mode(text: str) -> dict:
    """使用 JSON Mode"""
    msg = call_llm(
        messages=[
            {"role": "system", "content": "你是一个信息提取助手。请从用户输入中提取信息，返回 JSON 格式。"},
            {"role": "user", "content": f"从以下文本提取姓名、年龄、职业、城市、技能：\n\n{text}"},
        ],
        response_format={"type": "json_object"},
    )
    return json.loads(msg["content"])


# ═══════════════════════════════════════════════
# 方法3: Function Calling
# ═══════════════════════════════════════════════
EXTRACTION_SCHEMA = {
    "name": "extract_person_info",
    "description": "从文本中提取人物信息",
    "parameters": {
        "type": "object",
        "properties": {
            "name": {"type": "string", "description": "姓名"},
            "age": {"type": "integer", "description": "年龄"},
            "job": {"type": "string", "description": "职业"},
            "city": {"type": "string", "description": "所在城市"},
            "skills": {"type": "array", "items": {"type": "string"}, "description": "技能列表"},
            "email": {"type": "string", "description": "邮箱地址"},
        },
        "required": ["name", "age", "job"],
    },
}


def method3_function_calling(text: str) -> dict:
    """使用 Function Calling 强制输出"""
    msg = call_llm(
        messages=[
            {"role": "system", "content": "从用户输入中提取结构化信息。"},
            {"role": "user", "content": text},
        ],
        tools=[{"type": "function", "function": EXTRACTION_SCHEMA}],
        tool_choice={"type": "function", "function": {"name": "extract_person_info"}},
    )
    return json.loads(msg["tool_calls"][0]["function"]["arguments"])


# ═══════════════════════════════════════════════
# 主程序
# ═══════════════════════════════════════════════
def main():
    if not API_KEY:
        print("⚠️  未找到 API Key")
        return

    test_cases = [
        "我叫张三，28岁，软件工程师，精通Python和Java，邮箱zhangsan@email.com",
        "李四，35岁，北京，产品经理，5年互联网经验",
        "王五，42岁，深圳，技术总监，擅长团队管理和架构设计",
    ]

    methods = [
        ("📝 Prompt 约束", method1_prompt),
        ("🔧 JSON Mode", method2_json_mode),
        ("⚡ Function Calling", method3_function_calling),
    ]

    for text in test_cases:
        print("\n" + "=" * 60)
        print(f"📄 输入: {text}")
        print("=" * 60)

        for name, method in methods:
            try:
                result = method(text)
                # 验证是否为有效 JSON
                json.dumps(result, ensure_ascii=False)
                valid = "✅"
            except Exception as e:
                result = {"error": str(e)[:60]}
                valid = "❌"

            print(f"\n{valid} {name}:")
            print(f"   {json.dumps(result, ensure_ascii=False)}")


if __name__ == "__main__":
    main()

File: 02-进阶能力/code/day10/StructuredOutputDemo.java
// StructuredOutputDemo.java — 三种结构化输出方法对比（Java 版）
// 方法1: Prompt 约束 | 方法2: JSON Mode | 方法3: Function Calling
//
// 依赖: org.json (Maven: implementation 'org.json:json:20231013')
// 编译: javac -cp .:org.json-20231013.jar StructuredOutputDemo.java
// 运行: DEEPSEEK_API_KEY=xxx java -cp .:org.json-20231013.jar StructuredOutputDemo

import java.io.*;
import java.net.URI;
import java.net.http.*;
import java.net.URLEncoder;
import java.nio.charset.StandardCharsets;
import java.util.*;

public class StructuredOutputDemo {

    static final String API_KEY = System.getenv().getOrDefault("DEEPSEEK_API_KEY", "");
    static final String API_URL = "https://api.deepseek.com/v1/chat/completions";

    static HttpClient client = HttpClient.newHttpClient();

    // ── 调用 API ──
    static String callLLM(String jsonBody) throws Exception {
        HttpRequest req = HttpRequest.newBuilder()
            .uri(URI.create(API_URL))
            .header("Authorization", "Bearer " + API_KEY)
            .header("Content-Type", "application/json")
            .POST(HttpRequest.BodyPublishers.ofString(jsonBody))
            .timeout(java.time.Duration.ofSeconds(30))
            .build();
        HttpResponse<String> resp = client.send(req, HttpResponse.BodyHandlers.ofString());
        return resp.body();
    }

    static String extractContent(String response) {
        // 简易 JSON 解析：提取 content 字段
        int ci = response.indexOf("\"content\":\"");
        if (ci < 0) return null;
        int start = ci + 11;
        int end = response.indexOf("\"", start);
        return response.substring(start, end)
            .replace("\\n", "\n").replace("\\\"", "\"");
    }

    static String extractToolCallArgs(String response) {
        // 简易 JSON 解析：提取 tool_calls 中的 arguments
        int ai = response.indexOf("\"arguments\":\"");
        if (ai < 0) {
            // 有的 API 返回未转义的 JSON
            ai = response.indexOf("\"arguments\": ");
            if (ai < 0) return null;
            int start = response.indexOf("{", ai);
            int end = response.lastIndexOf("}") + 1;
            return response.substring(start, end);
        }
        int start = ai + 13;
        int end = response.indexOf("\"", start);
        return response.substring(start, end)
            .replace("\\n", "\n").replace("\\\"", "\"");
    }

    // ── 方法1: Prompt 约束 ──
    static String method1Prompt(String text) throws Exception {
        String prompt = "从以下文本中提取信息，只返回JSON（不要加任何其他文字）：\n\n"
            + "文本：" + text + "\n\n"
            + "要求的JSON格式：\n"
            + "{\"name\": \"姓名\", \"age\": \"年龄（数字）\", \"job\": \"职业\"}";

        String body = String.format(
            "{\"model\":\"deepseek-chat\",\"messages\":[{\"role\":\"user\",\"content\":\"%s\"}],\"temperature\":0}",
            prompt.replace("\\", "\\\\").replace("\"", "\\\"").replace("\n", "\\n")
        );
        String resp = callLLM(body);
        String content = extractContent(resp);
        if (content == null) return "解析失败";

        // 清理 markdown 包裹
        if (content.contains("```")) {
            String[] parts = content.split("```");
            for (String p : parts) {
                if (p.contains("{") && p.contains("}")) {
                    content = p;
                    break;
                }
            }
        }
        // 提取 JSON
        int start = content.indexOf("{");
        int end = content.lastIndexOf("}") + 1;
        return (start >= 0) ? content.substring(start, end) : "未找到 JSON";
    }

    // ── 方法2: JSON Mode ──
    static String method2JsonMode(String text) throws Exception {
        String body = String.format(
            "{\"model\":\"deepseek-chat\",\"messages\":[" +
            "{\"role\":\"system\",\"content\":\"你是一个信息提取助手。请从用户输入中提取信息，返回JSON格式。\"}," +
            "{\"role\":\"user\",\"content\":\"从以下文本提取姓名、年龄、职业：%s\"}]," +
            "\"response_format\":{\"type\":\"json_object\"},\"temperature\":0}",
            text.replace("\\", "\\\\").replace("\"", "\\\"").replace("\n", "\\n")
        );
        String resp = callLLM(body);
        String content = extractContent(resp);
        return (content != null) ? content : "解析失败";
    }

    // ── 方法3: Function Calling ──
    static String method3FunctionCalling(String text) throws Exception {
        String body = String.format(
            "{\"model\":\"deepseek-chat\",\"messages\":[" +
            "{\"role\":\"system\",\"content\":\"从用户输入中提取结构化信息。\"}," +
            "{\"role\":\"user\",\"content\":\"%s\"}]," +
            "\"tools\":[{\"type\":\"function\",\"function\":{\"name\":\"extract_person_info\"," +
            "\"description\":\"从文本中提取人物信息\"," +
            "\"parameters\":{\"type\":\"object\",\"properties\":{" +
            "\"name\":{\"type\":\"string\",\"description\":\"姓名\"}," +
            "\"age\":{\"type\":\"integer\",\"description\":\"年龄\"}," +
            "\"job\":{\"type\":\"string\",\"description\":\"职业\"}," +
            "\"city\":{\"type\":\"string\",\"description\":\"所在城市\"}" +
            "},\"required\":[\"name\",\"age\",\"job\"]}}}]," +
            "\"tool_choice\":{\"type\":\"function\",\"function\":{\"name\":\"extract_person_info\"}}," +
            "\"temperature\":0}",
            text.replace("\\", "\\\\").replace("\"", "\\\"").replace("\n", "\\n")
        );
        String resp = callLLM(body);
        String args = extractToolCallArgs(resp);
        return (args != null) ? args : "解析失败";
    }

    // ── 主函数 ──
    public static void main(String[] args) throws Exception {
        if (API_KEY.isEmpty()) {
            System.out.println("⚠️  请设置 DEEPSEEK_API_KEY");
            return;
        }

        String[] tests = {
            "我叫张三，28岁，软件工程师",
            "李四，35岁，北京，产品经理"
        };

        for (String text : tests) {
            System.out.println("=".repeat(55));
            System.out.println("📄 输入: " + text);
            System.out.println("=".repeat(55));

            System.out.println("\n📝 Prompt 约束:");
            System.out.println("  " + method1Prompt(text));

            System.out.println("\n🔧 JSON Mode:");
            System.out.println("  " + method2JsonMode(text));

            System.out.println("\n⚡ Function Calling:");
            System.out.println("  " + method3FunctionCalling(text));
            System.out.println();
        }
    }
}
